### PR TITLE
refactored Response for normalized handling independent of resource type

### LINKED ISSUE #842: Introduce Interfaces
### The problem

V3 of the SDK brings a Fluent interface to the end users but makes it difficult for them to build on top of it in a Fluent way.

Consider the following example use case where I might want to react to a 404 in a specific way, it might be by creating a new object, returning a different object or else.  

I currently would have to write something like this: 
```csharp
public static async Task<T> IfNotFound<T, T2>(
    this Task<T> sender, 
    Func<Task<T>> onMissing) 
    where T : Response<T2> 
{
    T result;

    try
    {
        result = await sender;
    }
    catch (CosmosException cosmosException)
    {
        if (cosmosException.StatusCode == HttpStatusCode.NotFound)
        {
            return await onMissing?.Invoke();
        }

        throw;
    }

    return result;
}
```
That could be used this way:
```csharp
var username = "testuser";
user = db.GetUser(username);

var userResponse = await user
    .ReadAsync()
    // Don't want to use Upsert, it would change the _ts
    .IfNotFound<UserResponse, UserProperties>(async () =>
    {
        return await db.CreateUserAsync(username);
    });

```
The necessity for `T2` comes from the fact that the SDK doesn't have any base class or interface allowing for a generic constraint that says "for a response" without knowing the return type up front.  I can't have a `Response` without knowing `T`.

What is missing is for a way to put a constraint around a `Response`.
### A possible solution
The SDK would benefit from something like the following: 

```csharp
public interface IResponse<T> : IResponseInformation
{
    T Resource { get; }
}

public interface IResponseInformation
{
    Headers Headers { get; }
    HttpStatusCode StatusCode { get; }
    double RequestCharge { get; }
    string ActivityId { get; }
    string ETag { get; }
    CosmosDiagnostics Diagnostics { get; }
}

// Response would only need to implement the new interface.
// This shouldn't break compatibility for anyone
public abstract class Response<T> : IResponse<T>
{
 ....
}
```
This would allow for:
```csharp
public static async Task<T> IfNotFound<T>(this Task<T> sender, 
    Func<Task<T>> onMissing)  where T : IResponseInformation
{
  ....
}
```
And in turn allow for:
```csharp
var username = "testuser";
user = db.GetUser(username);

var userResponse = await user
    .ReadAsync()
    .IfNotFound(async () => await db.CreateUserAsync(username));

```
I'm not clear if the low number of interfaces in the SDK is by choice or not, if that is the case, extracting a `Response` class out of `Response<T>` would accomplish the same as long as the non generic properties make it into the base class.

### LINKED ISSUE #1009: Microsoft.Azure.Cosmos.Response<T> should be refactored to enable response handling independent of resource type.
Microsoft.Azure.Cosmos.Response<T> should be refactored to in order to enable normalized handling of responses independent of resource type

```diff
namespace Microsoft.Azure.Cosmos
{
    using System.Net;

    /// <summary>
    /// The cosmos resource response class
    /// </summary>
-    public abstract class Response<T>
+    public abstract class Response
    {
        /// <summary>
        /// Gets the current <see cref="ResponseMessage"/> HTTP headers.
        /// </summary>
        public abstract Headers Headers { get; }

-        /// <summary>
-       /// The content of the response.
-        /// </summary>
-        public abstract T Resource { get; }
-
-        /// <summary>
-        /// Get Resource implicitly from <see cref="Response{T}"/>
-        /// </summary>
-        /// <param name="response">The Azure Cosmos DB service response.</param>
-        public static implicit operator T(Response<T> response)
-        {
-            return response.Resource;
-        }
-
        /// <summary>
        /// Gets the request completion status code from the Azure Cosmos DB service.
        /// </summary>
        /// <value>The request completion status code</value>
        public abstract HttpStatusCode StatusCode { get; }

        /// <summary>
        /// Gets the request charge for this request from the Azure Cosmos DB service.
        /// </summary>
        /// <value>
        /// The request charge measured in request units.
        /// </value>
        public abstract double RequestCharge { get; }

        /// <summary>
        /// Gets the activity ID for the request from the Azure Cosmos DB service.
        /// </summary>
        /// <value>
        /// The activity ID for the request.
        /// </value>
        public abstract string ActivityId { get; }

        /// <summary>
        /// Gets the entity tag associated with the resource from the Azure Cosmos DB service.
        /// </summary>
        /// <value>
        /// The entity tag associated with the resource.
        /// </value>
        /// <remarks>
        /// ETags are used for concurrency checking when updating resources. 
        /// </remarks>
        public abstract string ETag { get; }

        /// <summary>
        /// Gets the cosmos diagnostics information for the current request to Azure Cosmos DB service
        /// </summary>
        public abstract CosmosDiagnostics Diagnostics { get; }

        /// <summary>
        /// Gets the maximum size limit for this entity from the Azure Cosmos DB service.
        /// </summary>
        /// <value>
        /// The maximum size limit for this entity. Measured in kilobytes for document resources 
        /// and in counts for other resources.
        /// </value>
        /// <remarks>
        /// To get public access to the quota information do the following
        /// cosmosResponse.Headers.GetHeaderValue("x-ms-resource-quota")
        /// </remarks>
        internal abstract string MaxResourceQuota { get; }

        /// <summary>
        /// Gets the current size of this entity from the Azure Cosmos DB service.
        /// </summary>
        /// <value>
        /// The current size for this entity. Measured in kilobytes for document resources 
        /// and in counts for other resources.
        /// </value>
        /// <remarks>
        /// To get public access to the quota information do the following
        /// cosmosResponse.Headers.GetHeaderValue("x-ms-resource-usage")
        /// </remarks>
        internal abstract string CurrentResourceQuotaUsage { get; }
    }

+    /// <summary>
+    /// The cosmos resource response class
+    /// </summary>
+    public abstract class Response<T> : Response
+    {
+        /// <summary>
+        /// The content of the response.
+        /// </summary>
+        public abstract T Resource { get; }
+
+        /// <summary>
+        /// Get Resource implicitly from <see cref="Response{T}"/>
+        /// </summary>
+        /// <param name="response">The Azure Cosmos DB service response.</param>
+        public static implicit operator T(Response<T> response)
+        {
+            return response.Resource;
+        }
+    }
}
```

### PR DESCRIPTION
**Pull request by fuocor**
# Pull Request Template

## Description

Refactors Response<T> in order to support normalized handling of responses independent of ResourceType

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## Closing issues

Closes #1009 
Closes #842

### CODE REVIEW

**Comment by fuocor** [comment]
when running ```testbaseline.cmd /update``` it seems to have detected API changes in the ChangeFeedProcessor that were not made by me but in the master branch...

**Comment by j82w** [comment]
@ealsur can you investigate the change feed API changes?

File: Microsoft.Azure.Cosmos/src/Resource/Response.cs [hunk 1]
```diff
@@ -9,27 +9,13 @@ namespace Microsoft.Azure.Cosmos
     /// <summary>
     /// The cosmos resource response class
     /// </summary>
-    public abstract class Response<T>
+    public abstract class Response
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/src/Resource/Response.cs]: Please add a note in the changelog

File: Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DotNetSDKAPI.json [hunk 2]
```diff
@@ -543,6 +543,11 @@
           "Attributes": [],
           "MethodInfo": "Microsoft.Azure.Cosmos.ChangeFeedProcessorBuilder GetChangeFeedEstimatorBuilder(System.String, ChangesEstimationHandler, System.Nullable`1[System.TimeSpan])"
         },
+        "Microsoft.Azure.Cosmos.ChangeFeedProcessorBuilder GetChangeFeedProcessorBuilder(System.String, ChangesStreamHandler)": {
```
Comment (j82w) [inline Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DotNetSDKAPI.json]: The contract test failed on the check-in. Can you please sync to the latest and update the contract changes?

Comment (fuocor) [reply Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/DotNetSDKAPI.json]: synced and ran the contract tests and they passed

**Comment by ealsur** [comment]
Which are the changes that are detected? Are you doing a Clean before running that command? Sometimes if you switch branches, it still pickup stuff from the previous one.

**Comment by fuocor** [comment]
i ran ```Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/testbaseline.cmd /update```and got 
```
PS D:\git\ossiaco\azure-cosmos-dotnet-v3> Microsoft.Azure.Cosmos/tests/Microsoft.Azure.Cosmos.Tests/testbaseline.cmd /update
Settings:
TEST_BASELINE_PATH=bin\Release\netcoreapp2.0
TEST_OUTPUT_PATH=bin\Release\netcoreapp2.0
Calling Updating baselines...
xcopy "bin\Release\netcoreapp2.0\DotNetSDKAPIChanges.json" "D:\git\ossiaco\azure-cosmos-dotnet-v3\DotNetSDKAPI.json" /Q /Y
File not found - DotNetSDKAPIChanges.json
0 File(s) copied
Done.
```

the ContractChanges test is failing. What am IO doing wrong?

**Comment by ealsur** [comment]
Try to run it from within the folder containing the cmd file, not from another folder with the full path.

**Comment by j82w** [comment]
closes #842

File: Microsoft.Azure.Cosmos/src/Resource/Response.cs [hunk 3]
```diff
@@ -94,4 +80,24 @@ public abstract class Response<T>
         /// </remarks>
         internal abstract string CurrentResourceQuotaUsage { get; }
     }
+
+    /// <summary>
+    /// The cosmos resource response class
+    /// </summary>
+    public abstract class Response<T> : Response
```
Comment (kirankumarkolli) [inline Microsoft.Azure.Cosmos/src/Resource/Response.cs]: Ideally Response<T> : ResponseMessage makes sense. later is Disposable where as former is not.
In V4 base response is non-disposable so this aligns with V4 model. 

Can we pull this change in V4 instead?

**Comment by j82w** [comment]
This will be addressed in the v4 SDK which follows the [central SDKs model](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Azure.Core/src/ResponseOfT.cs).

**Pull request closed without merging**